source/develop/release-management/features/virt/dedicated-cpu.md

```diff
@@ -0,0 +1,201 @@
+    by this and any other VM (rendering all but one thread from a core unusable).
+  * *require-threads*: The host must have an SMT architecture. Each vCPU is
+    allocated on thread siblings and threads from single core can only be
+    assigned to this VM. If some threads remain unused these will be excluded
+    from use by any other VM.
+
+
+
+
+## Prerequisites
+
+There are no special prerequisites for this feature.
+
+
+## Security
+
+This feature is primarily designed for performance tuning and is not meant to
+be a security feature. It is possible for other VMs to share the dedicated CPUs
+for a short period of time when the VM is brought up (boot, after migration,
+restore from snapshots, etc.). This is not so much problem during boot because
+by the time OS is started the CPUs should be already evicted, but it could
+potentially provide a window of opportunity (for exploiting CPU
+vulnerabilities) after finished migration or when restoring VM from snapshot.
+
+
+## Limitations
+
+See the *Security* section.
+
+Currently, when a CPU is already dedicated to some VM, VDSM will not move the
+VM onto another CPU to make it free to another VM. This may be addressed in
+the future (see *Future Work* section), but for now this means:
+
+  * It may be impossible to start VMs with CPU pinning on that host because
+    the requested CPUs are already occupied. It is suggested to start VMs with
+    CPU pinning before any VMs with dedicated CPU policy.
+
+  * It may be impossible to start a VM with dedicated CPUs even though there
+    seems to be enough space in terms of number of used vs. free CPUs.
+
+
+## Changes to Engine and Vdsm
+
+The high level overview of the workflow in Engine and Vdsm is following:
+
+1.  When host switches to *UP* state engine queries it to find out which pCPUs
+    are dedicated through a policy or pinned. This is necessary for engine to
+    get a good picture about CPU usage so that it can schedule new VMs
+    properly. Although Engine is the one responsible for assigning the CPUs
+    and should know which CPUs are used by which VM, it may not remember the
+    state for recovering hosts or after Engine restart. Whether this is needed
+    at all probably depeonds on if the information about CPU allocations will
+    be stored in the database or not.
+
+2.  On VM start/migration/etc. Engine consults the policy for dedicating CPUs
+    and schedules the VM on host with enough free CPUs. Engine determines the
+    mapping between physical and virtual CPUs.
+
+3.  Engine produces domain XML that contains properly populated `<vcpupin>`
+    element (similarly to CPU pinning) and stores the policy name in VM
+    metadata.
+
+4.  Vdsm starts the VM and analyzes which CPUs were dedicated. Vdsm updates the
+    pool of shared CPUs and re-configures VMs without any CPU pinning or
+    dedicated CPUs to run on the new set of CPUs from shared pool.
```

Comment:
What with older Vdsm versions, which don't support this?

Reply:
The feature will be tied to some future cluster level so old VDSM should not be a problem. As for the opposite case (old engine and newer VDSM) we will treat all VMs without policy in metadata as shared unless they have explicit CPU pinning in domain XML.